lib/spatial_features/importers/file.rb

```diff
@@ -5,11 +5,21 @@ module Importers
     class File < SimpleDelegator
       INVALID_ARCHIVE = "Archive did not contain a .kml or .shp file. Supported formats are KMZ, KML, and zipped ArcGIS shapefiles.".freeze
 
-      def initialize(data, *args, **options)
-        # the current_file param is passed by `::create` after it has opened a zip
-        # archive and extracted the KML and SHP files
-        current_file = options.delete(:current_file)
+      def self.create_all(data, **options)
```

Comment:
This section seems awkward. For a zip containing shapefiles we end up unzipping it at least 2n+1 times where n is the number of zipped shapefiles. What if we just said by the time the file reaches `initialize` it should be unzipped and ready to go. If we operate under the assumption that the folder that contains the shapefile we pass the initializer must also contain the other required files (based on the name of the given shapefile), then we don't need to keep passing a zip file down to each successive level. That should even work with a zip with multiple shapefiles because the shapefile gem should match the other files based on the name of the shapefile. This way the importer instances are only ever dealing with a single file that's ready to go. Thoughts?

Reply:
I had the same thoughts so good we're discussing.  An earlier implementation did that, but it meant a few things:

- `Importer::Shapefile` would no longer accept a `ZIP` file, just a `SHP`
- we would need to tweak the SHX/DPF/PRF validation to look at sibling files instead of looking at the containing ZIP file (what the `rgeo` gem does)

I was concerned that removing the support for ZIP files in the Shapefile importer was going to cause problems so I walked it back.  I'm happy to pursue it though.

So our basic flow will become something like:

```
File::create_all
  download if necessary
  open zip if is zip
  each_recognized_file { process_with_child_importer }

Shapefile#init and Shapefile#create_all do the same thing:
  download if necessary
  only accepts a single SHP file
  fails if passed a ZIP
  performs validation against other files in `File.dirname(shapefile)`

KMLFile#init and KMLFile#create_all do the same thing:
  download if necessary
  only accepts a single KML file
```

The part that feels most uncomfortable to me is how we pass a file and then the importer looks around for sibling files.  Not sure why -- maybe it just feels less defined than I'm used to.  However, it's [exactly what rgeo-shapefile does](https://github.com/rgeo/rgeo-shapefile/blob/5230f3a10b969f59837422dc788cd2de5129fe12/lib/rgeo/shapefile/reader.rb#L179-L192) so we're already relying on that behaviour. \_(ツ)_/¯